torch/lib/c10d/reducer.cpp

```diff
@@ -391,7 +391,7 @@ void Reducer::copy_grad_to_bucket(
     auto wrapped = c10::scalar_to_tensor(double(1.) / divFactor_);
     wrapped.unsafeGetTensorImpl()->set_wrapped_number(true);
     // Divides while copying into the bucket view.
-    at::native::mul_out(bucket_view, grad, wrapped);
+    at::mul_out(bucket_view, grad, wrapped);
```

Comment:
So this was weird: I originally updated this to call `at::cpu::mul_out`, and I #included the `ATen/CPUFunctions.h` at the top of the file. That all compiled and linked without any errors, but when I tried to import torch from python I got some weird undefined symbols:

```
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/raid/hirsheybar/pytorch-port_structured_TI/torch/__init__.py", line 197, in <module>
    from torch._C import *
ImportError: /raid/hirsheybar/pytorch-port_structured_TI/torch/lib/libtorch_python.so: undefined symbol: _ZN2at3cpu7mul_outERNS_6TensorERKS1_S4_
```

I ran `c++filt` on that symbol to de-mangle it:
```
> c++filt _ZN2at3cpu7mul_outERNS_6TensorERKS1_S4_
at::cpu::mul_out(at::Tensor&, at::Tensor const&, at::Tensor const&)
```

That confirmed that it was coming from from my usage of `at::cpu::mul_out`, which I pinpointed to be from this file. I'm still not sure why that gave me an error, and it wasn't caught by the linker. The function implementation lives in `RegisterCPU.cpp`, and it's possible that the build workflow for `c10d` doesn't build that file or something.

Either way, I think it's probably fine to use `at::mul_out` here instead. It'll require another trip through the dispatcher, but that seems like less of a performance issue for c10d since it's distributed.

Reply:
This suggests c10d links against libtorch.so but not libtorch_cpu.so. Hmm, I wonder if we should have a TORCH_CPU_API macro. This workaround seems fine.